src/node_file_sync.cc

```diff
@@ -0,0 +1,350 @@
+        env,
+        permission::PermissionScope::kFileSystemWrite,
+        pathView,
+        Nothing<void>());
+  }
+  return JustVoid();
+}
+
+void BindingData::MemoryInfo(MemoryTracker* tracker) const {
+  // TODO(@anonrig): Implement this
+}
+
+BindingData::BindingData(Realm* realm, v8::Local<v8::Object> object)
+    : SnapshotableObject(realm, object, type_int) {
+  // TODO(@anonrig): Implement this
+}
+
+bool BindingData::PrepareForSerialization(v8::Local<v8::Context> context,
+                                          v8::SnapshotCreator* creator) {
+  // Return true because we need to maintain the reference to the binding from
+  // JS land.
+  return true;
+}
+
+InternalFieldInfoBase* BindingData::Serialize(int index) {
+  DCHECK_IS_SNAPSHOT_SLOT(index);
+  InternalFieldInfo* info =
+      InternalFieldInfoBase::New<InternalFieldInfo>(type());
+  return info;
+}
+
+void BindingData::Deserialize(v8::Local<v8::Context> context,
+                              v8::Local<v8::Object> holder,
+                              int index,
+                              InternalFieldInfoBase* info) {
+  DCHECK_IS_SNAPSHOT_SLOT(index);
+  v8::HandleScope scope(context->GetIsolate());
+  Realm* realm = Realm::GetCurrent(context);
+  BindingData* binding = realm->AddBindingData<BindingData>(holder);
+  CHECK_NOT_NULL(binding);
+}
+
+inline int32_t BindingData::CopyFileInternal(const std::string_view src,
+                                             const std::string_view dest,
+                                             int flags) {
+  uv_fs_t req;
+  auto make = OnScopeLeave([&req]() { uv_fs_req_cleanup(&req); });
+  FS_SYNC_TRACE_BEGIN(copyfile);
+  uv_fs_copyfile(nullptr, &req, src.data(), dest.data(), flags, nullptr);
+  FS_SYNC_TRACE_END(copyfile);
+
+  return int32_t(req.result);
+}
+
+void BindingData::CopyFile(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+
+  const int argc = args.Length();
+  CHECK_GE(argc, 3);
+
+  BufferValue src(isolate, args[0]);
+  CHECK_NOT_NULL(*src);
+  THROW_IF_INSUFFICIENT_PERMISSIONS(
+      env, permission::PermissionScope::kFileSystemRead, src.ToStringView());
+
+  BufferValue dest(isolate, args[1]);
+  CHECK_NOT_NULL(*dest);
+  THROW_IF_INSUFFICIENT_PERMISSIONS(
+      env, permission::PermissionScope::kFileSystemWrite, dest.ToStringView());
+
+  CHECK(args[2]->IsInt32());
+  const int flags = args[2].As<Int32>()->Value();
+
+  return args.GetReturnValue().Set(
+      CopyFileInternal(src.ToStringView(), dest.ToStringView(), flags));
+}
+
+int32_t BindingData::FastCopyFile(Local<Value> receiver,
+                                  const FastOneByteString& src,
+                                  const FastOneByteString& dest,
+                                  int32_t flags) {
+  // TODO(@anonrig): Add "THROW_IF_INSUFFICIENT_PERMISSIONS"
+  return CopyFileInternal(std::string_view(src.data, src.length),
```

Comment:
@joyeecheung I couldn't make the fast API work with the `bench-copyFileSync.js` benchmark. Somehow, V8 never get optimized enough to call this function. Do you have any idea?

PS: The benchmark uses `0` parameter for flags.

Reply:
Probably because the dest is not a flat one byte string.